### PR TITLE
New module: manage Foreman Global Parameters (remote_management/foreman/foreman_global_parameter)

### PR DESCRIPTION
**Pull request by Fobhep**
##### SUMMARY
We wrote an Ansible module to manage the Global Parameter Entitiy in Foreman with the nailgun library.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
foreman_global_parameter.py

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = 
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

### CODE REVIEW

File: lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        set(),
+        {'search': '%s="%s"' % (key, kwargs[key]) for key in kwargs}
+    )
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            server_url=dict(required=True),
+            username=dict(required=True, no_log=True),
```
Comment (gundalow) [inline lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py]: Username shouldn't need `no_log`

File: lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+            password=dict(required=True, no_log=True),
+            verify_ssl=dict(required=False, type='bool', default=False),
+            name=dict(required=True, no_log=False),
```
Comment (gundalow) [inline lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py]: Please remove `no_log=False` as it's the default

File: lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            value=dict(required=False, no_log=False),
```
Comment (gundalow) [inline lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py]: Please remove `no_log=False` as it's the default

File: lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+            state=dict(required=True, no_log=False),
```
Comment (gundalow) [inline lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py]: Please remove `no_log=False` as it's the default

File: lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+            - Username on Foreman server
+        required: true
+    password:
+        description:
+            - Password for user accessing Foreman server
+        required: true
+    verify_ssl:
+        description:
+            - Verify SSL of the Foreman server
+        required: false
```
Comment (gundalow) [inline lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py]: `default: true`

File: lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    )
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            server_url=dict(required=True),
+            username=dict(required=True, no_log=True),
+            password=dict(required=True, no_log=True),
+            verify_ssl=dict(required=False, type='bool', default=False),
```
Comment (gundalow) [inline lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py]: Please change the default to true (assuming this is what other Forman modules do)

File: lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+            name=dict(required=True, no_log=False),
+            value=dict(required=False, no_log=False),
+            state=dict(required=True, no_log=False),
+        ),
+        supports_check_mode=True
+    )
+    if not HAS_NAILGUN_PACKAGE:
+        module.fail_json(
+            msg="""Missing required nailgun module (check docs or install
```
Comment (gundalow) [inline lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py]: Missing closing `)`

File: lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    changed = False
+    try:
+        create_Server(server_url, auth, verify_ssl)
+        param = find_Entity(CommonParameter, name=name)
+    except Exception as e:
+        module.fail_json(msg="Failed to connect to Foreman server: %s " % e)
+    try:
+        if state == 'present':
+            if not value:
+                module.fail_json(msg='value is required when state=present')
```
Comment (gundalow) [inline lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py]: There are built in helpers for this,

See the following draft example https://github.com/gundalow/ansible/blob/docs-argspec/docs/docsite/rst/dev_guide/developing_modules_general.rst#main-and-ansiblemodule-argument-spec

File: lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+                    changed = True
+            else:
+                create_CommonParameter(name, value)
+                changed = True
+        elif state == 'absent':
+            if len(param) == 1:
+                param[0].delete()
+                changed = True
+        else:
+            module.fail_json(msg='state can only be present or absent')
```
Comment (gundalow) [inline lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py]: Please remove and add the suggested line in the arg spec

File: lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py (hunk 4 above)
Comment (gundalow) [inline lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py]: Please add `, choices=['present', 'absent']`

File: lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+            - Name of the Global Parameter
+        required: true
+    value:
+        description:
+            - Value of the Global Parameter
+        required: false
+    state:
+        description:
+            - State of the Global Parameter
+        required: false
```
Comment (gundalow) [inline lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py]: Please add `choices:` here

http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html

**Review by gundalow — CHANGES_REQUESTED** [review]

**Comment by gundalow** [comment]
@Fobhep Thanks for the PR. I believe this is your first, ace :)
The following page has details of regarding how to document your module, which may be of use https://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html

@ehelms I believe you created the other Foreman modules, it would be good to get your input on this new module.

**Comment by Fobhep** [comment]
@gundalow thx for the review. 
Hehe, yes it is indeed my first, thx :)

I wrote this basically to try and get a feeling for nailgun and foreman entities.
There could come more at a later point, regarding other foreman entitites. ;)

**Review by gundalow — APPROVED** [review]

**Comment by ehelms** [comment]
Sorry - just now seeing this and taking a look.

File: lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+    nailgun.entity_mixins.EntityDeleteMixin,
+    nailgun.entity_mixins.EntityReadMixin,
+    nailgun.entity_mixins.EntitySearchMixin,
+    nailgun.entity_mixins.EntityUpdateMixin
+):
+
+    pass
+
+
+def create_CommonParameter(name, value):
```
Comment (ehelms) [inline lib/ansible/modules/remote_management/foreman/foreman_global_parameter.py]: This mixture of snakecase and camcelcase for method names seems a bit strange and divergent from typical python best practices. A bit of syntactic nitpick but it is nice to follow standard practice.

**Comment by ehelms** [comment]
Minor syntactic nitpick but the strategy looks good and is in line with how I want to re-factor the katello module.

**Comment by Fobhep** [comment]
@ehelms thank your very much for your review and your suggestion. I will change the synaptics of my own code into a more standard practice conform way.

**Comment by Fobhep** [comment]
ok, seems to be fine now :)

**Comment by maxamillion** [comment]
@Fobhep apologies for the lag time on this. If you're still interested in contributing this, the code looks good. We'd just need to update the `version_added` field and add some tests. Thank you!

**Comment by ekohl** [comment]
wontfix because I think https://github.com/theforeman/foreman-ansible-modules is a better place for this (and already contains an updated version of this PR).

**Pull request closed without merging**